### ISSUE TITLE
restic backup fails to save blobs and renew locks during backup

### ISSUE THREAD

**Issue by borkd** [restic/restic#3265]
Output of `restic version`
--------------------------
restic 0.11.0 compiled with go1.15.3 on linux/amd64

How did you run restic exactly?
-------------------------------

```
$ restic backup -x --tag tag1 --tag tag2 --tag tag3 dir1 dir2 dirprefix*
repository 41367ccf opened successfully, password is correct
created new cache in redacted/.cache/restic
Save(<data/f6d39f650f>) returned error, retrying after 738.458435ms: client.PutObject: Your socket connection to the server was not read from or written to within the timeout period.
Save(<data/c115994973>) returned error, retrying after 287.145499ms: client.PutObject: Your socket connection to the server was not read from or written to within the timeout period.
Save(<data/5a0b25cb1e>) returned error, retrying after 361.144708ms: client.PutObject: Your socket connection to the server was not read from or written to within the timeout period.
Save(<data/17b2c54475>) returned error, retrying after 590.539156ms: client.PutObject: Your socket connection to the server was not read from or written to within the timeout period.
Save(<data/f354e5187b>) returned error, retrying after 370.757544ms: client.PutObject: Your socket connection to the server was not read from or written to within the timeout period.
Save(<data/3f56ada95c>) returned error, retrying after 615.115739ms: client.PutObject: Your socket connection to the server was not read from or written to within the timeout period.
Save(<data/4e129fb137>) returned error, retrying after 341.462458ms: client.PutObject: Your socket connection to the server was not read from or written to within the timeout period.
Save(<data/93872cbf61>) returned error, retrying after 591.326744ms: client.PutObject: Your socket connection to the server was not read from or written to within the timeout period.
Save(<data/e76945d4a8>) returned error, retrying after 739.464678ms: client.PutObject: Your socket connection to the server was not read from or written to within the timeout period.
Save(<data/47c56f37ec>) returned error, retrying after 283.564177ms: client.PutObject: Your socket connection to the server was not read from or written to within the timeout period.
Save(<data/7444d335c8>) returned error, retrying after 575.230153ms: client.PutObject: Your socket connection to the server was not read from or written to within the timeout period.
Save(<data/44d7613d65>) returned error, retrying after 514.307064ms: client.PutObject: Your socket connection to the server was not read from or written to within the timeout period.
Save(<data/17d3d5a19b>) returned error, retrying after 619.667246ms: client.PutObject: Your socket connection to the server was not read from or written to within the timeout period.
Save(<data/2ac1628a6f>) returned error, retrying after 713.381054ms: client.PutObject: Your socket connection to the server was not read from or written to within the timeout period.
Save(<data/b9d73f1a2e>) returned error, retrying after 575.126373ms: client.PutObject: Your socket connection to the server was not read from or written to within the timeout period.
Save(<data/f1893a13a9>) returned error, retrying after 491.58676ms: client.PutObject: Your socket connection to the server was not read from or written to within the timeout period.
Save(<data/80c4416bb1>) returned error, retrying after 391.203249ms: client.PutObject: Your socket connection to the server was not read from or written to within the timeout period.
Save(<data/2f94e1d972>) returned error, retrying after 570.57463ms: client.PutObject: Your socket connection to the server was not read from or written to within the timeout period.
Save(<data/85af09c76c>) returned error, retrying after 366.273224ms: client.PutObject: Your socket connection to the server was not read from or written to within the timeout period.
[122:02:21] 83.05%  11565236 files 7.909 TiB, total 14151160 files 9.524 TiB, 0 errors ETA 24:54:31

```

What backend/server/service did you use to store the repository?
----------------------------------------------------------------

s3 swiftstack (less than 2ms away)

Expected behavior
-----------------

Reliable and predictable backup/upload/restore behavior or a loud failure as early as possible. 

Ideally `restic backup` should retry saving the blob until it succeeds, possibly followed by a read request to verify that each of the small fraction of problematic blobs exists on the backend. User should be able to specify the cound and duration of upload retries. Once the retry counter is reached backup should exit right away with a non-zero exit code.

Actual behavior
---------------

Warning message is logged, as above, and backup continues, possibly for days, while the blob has never reached the backend.

```
minio-mc find backend/bucket | grep data/17d3d5a19b
``` 
returns no entries

Steps to reproduce the behavior
-------------------------------

- start doing backups of sufficient volume
- wait for error message or simulate network congestion/disruptions between restic and the backend
- watch restic log the error and continue as if nothing has happened, possibly for days

Do you have any idea what may have caused this?
-----------------------------------------------

insufficient error handling?

Do you have an idea how to solve the issue?
-------------------------------------------

Pause/queue processing on upload errors until they are resolved and validated or fail the backup right there and then according to user specified acceptable retry count / time.

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------

My first impression is this bug report. There is hope, but not yet

**Comment by borkd** [comment]
#3263 "Data blobs seem to be missing, aborting prune to prevent further data loss! " on another repository appears to be related result of this lack of failure handling

**Comment by borkd** [comment]
I suspect similar failures are allowed to happen during lock creation/reresh/removal for long running jobs when network bandwidth between restic and the backend is only a fraction of disk bandwidth on either side. One of the repositories which was created with this version about a week ago started reporting spurious lock issues

**Comment by borkd** [comment]
After a couple of more hours of this multi-day backup session I am seeing the message indicating lock renewal is affected by this same problem:

```
[...]
Save(<lock/c5ea81c926>) returned error, retrying after 352.204325ms: client.PutObject: Put "https://redacted/bucket/locks/c5ea81c9268333e630c2cf8106b26fc15e0ee06c51c828ce800f53f9fb1aca33": dial tcp redacted:443: i/o timeout
Save(<data/de9250a8aa>) returned error, retrying after 392.621188ms: client.PutObject: Your socket connection to the server was not read from or written to within the timeout period.
Save(<data/dd7ff8f1ff>) returned error, retrying after 658.500365ms: client.PutObject: Your socket connection to the server was not read from or written to within the timeout period.
[134:42:00] 88.97%  13904763 files 8.474 TiB, total 14151160 files 9.524 TiB, 0 errors ETA 16:41:30
[... backup continues ...]
```

**Comment by borkd** [comment]
**0 errors** in the above status is a proof current failure handling protocol is cheerfully clueless about the severity of the problem

**Comment by MichaelEischer** [comment]
> Ideally `restic backup` should retry saving the blob until it succeeds, possibly followed by a read request to verify that each of the small fraction of problematic blobs exists on the backend. User should be able to specify the cound and duration of upload retries. Once the retry counter is reached backup should exit right away with a non-zero exit code.

The log snippet in your initial report sort of shows that restic retried the request and seemed to be happy with the result. As far as I can see the log only contains reports that the first try to upload some of the pack files did fail. restic retries each failed upload 10 times for a total of at most 15 minutes. If any upload fails permanently then the backup process will abort. So your log snippet shows that _the storage backend has reported the retried uploads to be successful_.

The only part of your suggestion which restic currently doesn't implement is trying to access the file after it was uploaded. But I doubt that adding such a check would reliably solve the problem that a file becomes missing for some reason. Just to repeat the important part here: restic was told by the backend that the upload was successful because otherwise it would have canceled the backup run. So the real question here is: why is the file no longer there?

I've seen the "Your socket connection to the server was not read from or written to within the timeout period." error only once in another old issue. According to 
https://aws.amazon.com/premiumsupport/knowledge-center/s3-socket-connection-timeout-error/ it could be related to network problems as apparently a upload would have to be stuck for a few dozen seconds. But that doesn't really fit your 2ms latency number...

> Warning message is logged, as above, and backup continues, possibly for days, while the blob has never reached the backend.
> 
> ```
> minio-mc find backend/bucket | grep data/17d3d5a19b
> ```
> 
> returns no entries

You are looking at the wrong path. The file should be stored in `data/17/17d3d5a19b[...]` (the dots represent the remaining 54 characters of the filename).

> **0 errors** in the above status is a proof current failure handling protocol is cheerfully clueless about the severity of the problem

0 errors refers to problems with reading the files that should be backed up, so there's nothing wrong here.

**Comment by fd0** [comment]
> 0 errors in the above status is a proof current failure handling protocol is cheerfully clueless about the severity of the problem

Your text reads to me as very arrogant. Can you please tone it down? Thanks!

> Ideally restic backup should retry saving the blob until it succeeds, possibly followed by a read request to verify that each of the small fraction of problematic blobs exists on the backend. User should be able to specify the cound and duration of upload retries. Once the retry counter is reached backup should exit right away with a non-zero exit code.

This is **exactly** what restic does :)

> Warning message is logged, as above, and backup continues, possibly for days, while the blob has never reached the backend.

As @MichaelEischer already wrote: your test is wrong, the file is stored in a subdir called `data/17`.

> Save(<data/f6d39f650f>) returned error, retrying after 738.458435ms: client.PutObject: Your socket connection to the server was not read from or written to within the timeout period.
> [...]
> Save(<lock/c5ea81c926>) returned error, retrying after 352.204325ms: client.PutObject: Put "https://redacted/bucket/locks/c5ea81c9268333e630c2cf8106b26fc15e0ee06c51c828ce800f53f9fb1aca33": dial tcp redacted:443: i/o timeout
> Save(<data/de9250a8aa>) returned error, retrying after 392.621188ms: client.PutObject: Your socket connection to the server was not read from or written to within the timeout period.

What you're seeing here are warnings that restic did not succeed to save a file on the first try. The file may be one stored in a subdir of `data` (`Save(<data/f6d39f650f>) returned error[...]`) or `locks` (`Save(<lock/c5ea81c926>) returned error[...]`). In general, restic's policy is to report errors and try to mitigate them, which is what it does: The upload is retried (with a backoff) and succeeds, otherwise you would have seen another error message for the same file name. If even the fifth try does not succeed, restic exits with an error code. From the log excerpt you pasted I can't see two error messages for the same file, so I conclude that the upload succeeded on the second try each time.

> Save(<data/de9250a8aa>) returned error, retrying after 392.621188ms: client.PutObject: Your socket connection to the server was not read from or written to within the timeout period.

This error message tells me the underlying reason for the errors: restic keeps a pool of connections to the backend, which are either reused (when there's demand) or closed after a timeout of 90 seconds. It looks like the backend server or an intermediate proxy has a lower timeout for idle connections and closes them, which restic only notices when it tries to reuse a connection. In this case an error is returned and restic tries again after a short timeout. The second upload succeeds, so it does not print anything else and continues the backup.

I have some ideas what you can try:
 * Raise the timeout for idle connections
 * If your backend supports HTTP2 you you can try restic compiled from master (you can find pre-compiled binaries [here](https://beta.restic.net/restic-v0.11.0-375-g9985368d/)) which enables HTTP2 (we accidentally disabled it some time ago). That way you'll only see a single connection over which all requests are multiplexed, therefore you don't have the situation that a connection is idle.

I think restic does the right thing here, so I'm closing this issue for now. Please feel free to add further comments so we can reopen it if there's anything we need to do. Thanks!

**Comment by borkd** [comment]
Apologies, indeed I would not phrase the  0 errors comment the same way if I were doing it now. I did not mean the to sound arrogant, but rather raise the issue of user experience. If errors and/or warning conditions are encountered with any part of the process (reading/sending/writing) it makes more sense to print counters showing the entire picture, not just errors related to reading data. Giving users full stats of the backup operation, instead of only read-side errors may give users false sense of security, especially with a number of existing open issues complaining about missing blobs.

Another idea, which could help people troubleshoot their backups and provide useful feedback would be to enable tracing/debugging on the fly when a problem is encountered, and disabbling it after restic believes all is well again. Building a debug binary and running entire session with it is not practical as it will likely slow down the process enough to alter the overall situation. Assuming restic always does the right thing, it would then point out weak spots in interactions with backends which then could be acted upon. Either way this would be beneficial.

As for closing this case, your call. I am seeing errors reported above and trying to pinpoint what is breaking and where

**Comment by borkd** [comment]
@fd0 - practical question related to beta build with HTTP/2 support. Initial backup job that is missing data blobs for tiny fraction of overall data stored in the bucket ran for almost a week. I have downloaded the contents of the bucket for tests as a local repo and the blobs are missing from the `data/XY/XYhex`.

Will a subsequent runs of `restic backup` rely on the existing indexes and skip over source files for which data is assumed to exist in the backend but is actually missing? Or is there a way for `restic check` to mark problematic blobs as dirty in the index and forcing future backup sessions to process and upload just the respective bits?

**Comment by fd0** [comment]
> Initial backup job that is missing data blobs for tiny fraction of overall data stored in the bucket ran for almost a week.

Is that the backup run quoted in this issue? The errors you reported so far looked as if restic was successful on the second upload.

What does `restic check` report exactly?

> Initial backup job that is missing data blobs for tiny fraction of overall data stored in the bucket ran for almost a week. I have downloaded the contents of the bucket for tests as a local repo and the blobs are missing from the `data/XY/XYhex`.

In restic's vocabulary that's a "pack" or "pack file" and not a "blob" (the design document contains a [section about terminology](https://github.com/restic/restic/blob/master/doc/design.rst#terminology)).

I'm surprised that restic managed to complete a backup thinking it uploaded all the data successfully but apparently files are missing, that's odd.

> Will a subsequent runs of restic backup rely on the existing indexes and skip over source files for which data is assumed to exist in the backend but is actually missing?

It will rely on the existing indexes. If whole files are missing you can build a new index using `restic rebuild-index` which won't contain the missing files any more. The next run of `restic backup` will upload all data that's missing.

**Comment by MichaelEischer** [comment]
> Giving users only read-side errors of the backup operation, instead of full stats, may give users false sense of security, especially with a number of existing open issues complaining about missing blobs.

restic considers any write error which can't be resolved by retries as fatal. The log shows that from restic's point of view the following has happened: Save(<data/f6d39f650f>) -> Timeout. Then restic tries to cleanup possible leftover data Remove(<data/f6d39f650f>) -> Probably "not found". And finally the actual retry: Save(<data/f6d39f650f>) -> Success. At which point everything looks alright.

> If whole files are missing you can build a new index using `restic rebuild-index` which won't contain the missing files any more

Rebuilding the index will be much faster when using a beta build. But please run `check` before doing so.

**Comment by MichaelEischer** [comment]
@borkd Do you have access to logs on the storage server? It would be interesting which requests were issued by restic e.g. for `data/f6/f6d39f650f`.

**Comment by borkd** [comment]
> restic considers any write error which can't be resolved by retries as fatal. 

@MichaelEischer such approach makes perfect sense for software where consistency should come first. Software breaks, backends lie, networks have problems, that's why people do backups.

Being backup software, my suggestion is to consider printing retry/warning counters in addition to read-side counter, and
enabling on-demand logging/tracing whenever retries are activated, and to follow up with a stat or head on the object as verification when the recovery succeeds before the count of 10. If such a trace is produced and the exit was 0, a good way to build trust would be to suggest user runs a check on the repo and if anything is amiss ask them to open an issue and include the trace. Surely a non-zero effort to have this implemented, but it would provide long term benefit to the users, and provide timely and invaluable feedback for both existing and future backends as they evolve new bugs.

> Is that the backup run quoted in this issue? The errors you reported so far looked as if restic was successful on the second upload.

@fd - Yes. I've had a number of long backup jobs running simultaneously, most to separate repositories but a few into a shared repo. Some shorter job running during the same time window did exit with non-zero exit code after reporting errors, so I had a heads up I needed to run them again. But some finished cleanly, despite were missing individual blobs. This prompted me to poke around the longest and most resource intensive backups, which spat the occasional "retrying" as noted above during congestion events, but eventually also finished with return code of 0.

I have downloaded the contents of the bucket into a minio instance on local network and started `backup --force` against it to ensure I do have a complete backup. That is still running, I have not seen any retries and don't expect to see any as there no chance of congestion. Once finished I will run a check.

> Do you have access to logs on the storage server? It would be interesting which requests were issued by restic e.g. for data/f6/f6d39f650f.

@MichaelEischer - I am working on getting as much info from the backend operator as I can. When problems were first discovered I have received a verbal "all seems good on our side". I do not know the extent of logging and retention they have in place but I will try to get more info this week. Having an automatic trace generated and in hand would be golden.. especially if the backend lied, or something else was happening with the network.

@MichaelEischer - how "safe" is the *beta* to use? Is it potentially a one way street? Some of the data in question has a fixed TTL and I may not be able to rerun on the same dataset over and over again. I would definitely prefer faster operation and HTTP/2, but not if it leaves me without an option to go back to "stable" tree

**Comment by fd** [comment]
fwd: @fd0

**Comment by fd0** [comment]
@fd thanks a lot! :)

**Comment by borkd** [comment]
@MichaelEischer and @fd0 - I was not successful in getting any logs or specific information from the backend operators besides a vague mention of some disks having problems, and a question whether I attempted to read/access the problematic object after creation. While such failures are rare a selective access test for retried objects may not be a bad idea, and would augment the existing repo check functionality.

Here is the actual output from check done with the beta on one of the affected repositories:

```
create exclusive lock for repository
load indexes
check all packs
pack 01bfcb35: does not exist
pack 2b215f11: does not exist
pack f4963f90: does not exist
pack f96f5742: does not exist
pack 4a28ebad: does not exist
pack 40b082ea: does not exist
pack 06da7da9: not referenced in any index
[ ... ]
276 additional files were found in the repo, which likely contain duplicate data.
You can run `restic prune` to correct this.
check snapshots, trees and blobs
[0:18] 100.00%  353 / 353 snapshots

Fatal: repository contains errors
```

Do you have any suggestions before rebuild-index?

**Comment by fd0** [comment]
> I would definitely prefer faster operation and HTTP/2, but not if it leaves me without an option to go back to "stable" tree

I can understand that. Despite any bugs we are not aware of, restic compiled from master is fully compatible with 0.11.0 and you should be able to go back and forth between the versions.

**Comment by borkd** [comment]
> > If whole files are missing you can build a new index using `restic rebuild-index` which won't contain the missing files any more
> 
> Rebuilding the index will be much faster when using a beta build. But please run `check` before doing so.

I have since used the beta for faster checks and backups using HTTP/2. Check on another repository (10.623T, Objects: 2252856) reveals 59 missing packs:

```
using temporary cache in /tmp/restic-check-cache-780555952
created new cache in /tmp/restic-check-cache-780555952
create exclusive lock for repository
load indexes
check all packs

17348 additional files were found in the repo, which likely contain duplicate data.
You can run `restic prune` to correct this.
check snapshots, trees and blobs
[0:11] 100.00%  80 / 80 snapshots

read group #1 of 8692 data packs (out of total 2235087 packs in 256 groups)
[5:34] 100.00%  8692 / 8692 packs
pack cf2a1268: does not exist
pack 2f1f175d: does not exist
pack bbfbd867: does not exist
pack b76ec255: does not exist
pack c7c9d871: does not exist
pack a4612e36: does not exist
pack 182c1476: does not exist
pack 93059618: does not exist
pack 8a150dad: does not exist
pack 3d8f74ad: does not exist
pack 4b42702f: does not exist
pack a1ab775a: does not exist
pack 5b5c14db: does not exist
pack 94f4e450: does not exist
pack a6494d7f: does not exist
pack 533822c6: does not exist
pack e61a0960: does not exist
pack f9e5d55c: does not exist
pack edb8df28: does not exist
pack e7d57c83: does not exist
pack dc7b6352: does not exist
pack 72964f96: does not exist
pack 16ec1bbd: does not exist
pack 5171dec9: does not exist
pack 250bff63: does not exist
pack 44b5bfe5: does not exist
pack 4a7f9596: does not exist
pack f23f41fe: does not exist
pack af49f724: does not exist
pack 986aadfc: does not exist
pack 213449eb: does not exist
pack 5be6d03f: does not exist
pack 17fd00bb: does not exist
pack 8dfa0e0b: does not exist
pack 6a615b62: does not exist
pack eaaff233: does not exist
pack 536fb84c: does not exist
pack d603a32e: does not exist
pack f862465a: does not exist
pack 3588f1c6: does not exist
pack 6837d267: does not exist
pack 4b44bc42: does not exist
pack 7a7a43bb: does not exist
pack e47d944e: does not exist
pack e127b246: does not exist
pack f7e608ad: does not exist
pack 32892606: does not exist
pack 2776af7d: does not exist
pack 82fdf64b: does not exist
pack a50f4968: does not exist
pack d82d4605: does not exist
pack 14c27184: does not exist
pack 4b2b6216: does not exist
pack 216bd7b7: does not exist
pack 6026a7e0: does not exist
pack 7735fa50: does not exist
pack 005d57cf: does not exist
pack 3e500fc1: does not exist
pack d45416e7: does not exist
pack eaf4b664: not referenced in any index
pack 4745a714: not referenced in any index
pack 5b7fb049: not referenced in any index
pack be1169f3: not referenced in any index
pack 077289d0: not referenced in any index
[...]
17348 additional files were found in the repo, which likely contain duplicate data.
You can run `restic prune` to correct this.
check snapshots, trees and blobs
[0:11] 100.00%  80 / 80 snapshots

read group #1 of 8692 data packs (out of total 2235087 packs in 256 groups)
[5:34] 100.00%  8692 / 8692 packs

Fatal: repository contains errors
```

@MichaelEischer , @fd0 - I went through the existing documentation looking for troubleshooting/recovery steps, then forums, and open PRs but this topic appears to have rather sparse coverage. Considering I want to treat all repositories as append only as much as possible -  ie. entire repo rotation over time, rather than pruning, what is the recommended sequence of commands to troubleshoot and "repair" a damaged repo? I can see what worked or did not work, and submit a PR for the documentation once I am able to get a passing `restic check` for currently affected repositories.

**Comment by MichaelEischer** [comment]
Repairing a damaged repository usually

**Comment by MichaelEischer** [comment]
Repairing a damaged repository should preferably be done after understanding how the repository was damaged and if necessary fixing the root cause. Did that repository also have the errors reported from PutObject?

Depending on the exact type of damage, the steps to recover from it need some variation. In case of your repository with missing files probably the following should work:

- If possible create a backup of the repository's index folder, that could be useful to investigate the damage. Then run `restic rebuild-index` (either use the beta version or restic 0.12.0)
- Then it's probably a good idea to run all backup tasks again, these could pick up some blobs again which are now considered missing. (With restic before 0.10.0 you have to add the `--force` flag).
- The run `restic check` to learn which blobs are still missing.
- With `restic find --blob <blob-id> ...` or `restic find --tree <tree-id>` you can search for the damaged snapshots and remove those.

**Comment by borkd** [comment]
I have been using beta for HTTP/2 and speed since it was mentioned, and switched to the 12 (latest release) when it came out. 

Does it make sense to first try locating files represented by missing packs and backup them again, if possible? Or is there a chance existing indexes, prior to rebuild, may not have the full view of what was damaged? I expect this would be orders of magnitude faster than redoing tens of millions of files, many of which may have since changed or may now be gone. 

As to the root cause, I agree it would be best to learn what exactly went on, but that seems less and less likely. I have only fragmented bits and pieces of information that indicate the storage backend experienced a relatively rare but not unexpected cascade of problems just as I was running backups against it. When problems occurred I had between 5-10 simultaneous backups running against buckets with separate repositories. Some jobs aborted, which brought my attention to the issue, but others went on and eventually succeeded without error. Checks on faster jobs which finished yielded errors. Seeing retries and timeouts in the still running job I went on to poke around the backend bucket and found that objects restic was retrying and eventually moved on were nowhere to be found. This issue was opened. Vendor asked if I tried to access/verify the objects after creation was running into issues and that's about it. Since then I've pulled data from most buckets with problems to local machines and am trying to repair the contents before sending the fixed repos back. 

An option for, or a mandatory automatic selective verification of new/modified object's metadata when retries/timeouts are triggered would be very handy.

**Comment by MichaelEischer** [comment]
> Does it make sense to first try locating files represented by missing packs and backup them again, if possible? Or is there a chance existing indexes, prior to rebuild, may not have the full view of what was damaged? I expect this would be orders of magnitude faster than redoing tens of millions of files, many of which may have since changed or may now be gone.

Running `backup` again does not help as long as the index references missing pack files. `backup` assumes that everything contained in the index still exists and won't backup missing parts again. So the first step to repair a repository _must_ always be to run `rebuild-index`. For future reference: Note that it is a good idea to investigate the repository damage before rebuilding the index as that might obscure what went wrong.

> An option for, or a mandatory automatic selective verification of new/modified object's metadata when retries/timeouts are triggered would be very handy.

It wouldn't be too hard to implement that. The main reason why I'm not yet convinced of such a verification is that it's effectiveness depends a lot on what the actual problem was. For the retried file upload, restic effectively issued the following operations: `save(data/...)=error`, `delete(data/...)=error?` followed by `save(data/...)=ok`. Now calling `stat(data/...)` for verification should successfully be able to find the file (assuming certain types of consistency, which all(?) backends should be able to satisfy). It is rather unclear to me how a storage backend would lose files after confirming a successful upload. Possible ways would be a delayed cleanup after the failed initial save, a delayed application of the delete operation or a completely broken save operation. The first two cases would probably only take effect after the sanity check. And the last case would affect each save operation, not just the retried ones, and thus would silently break the repository.

**Comment by borkd** [comment]
When it comes to a root cause, I am interested in finding that out through out via a simulated failing environment when I get some spare cycles. No idea how or if this will be successful for the exact case I have ran into. Having some instrumentation in restic to show that fault handling is activated without running entire session in debug mode would help immensely in finding fragile cases where deduped user data may be exposed to extra risk

**Comment by borkd** [comment]
Breaking the repo is definitely not what we are after, same goes for slowing backups down and making them more expensive with stats after each upload. For figuring out what the best approach is I would suggest just emitting a time limited log messages when error handling is activated. Comparing time stamps from restic job log with time stamps generated in the fault-injection backend should yield some insight on fragile situations

**Comment by borkd** [comment]
@MichaelEischer, @fd0

This is a log excerpt from a current long running backup job:

```
[...]
repository 41367ccf opened successfully, password is correct
no parent snapshot found, will read all files
Save(<data/17baae2f3b>) returned error, retrying after 561.584179ms: wrote 0 bytes instead of the expected 4317764 bytes
Save(<data/e2f5aeb8ce>) returned error, retrying after 649.286798ms: wrote 0 bytes instead of the expected 6887782 bytes
Save(<data/c69614f9b8>) returned error, retrying after 502.453954ms: wrote 0 bytes instead of the expected 6515973 bytes
Save(<data/ea7b1133b0>) returned error, retrying after 417.451519ms: wrote 0 bytes instead of the expected 6805934 bytes
Save(<data/0e873759aa>) returned error, retrying after 529.29694ms: wrote 0 bytes instead of the expected 5023131 bytes
Save(<data/a4b0010091>) returned error, retrying after 377.89376ms: wrote 0 bytes instead of the expected 6638624 bytes
Save(<data/635353a728>) returned error, retrying after 631.124104ms: wrote 0 bytes instead of the expected 7787537 bytes
Save(<data/7ad3c44a1d>) returned error, retrying after 548.316808ms: wrote 0 bytes instead of the expected 6368056 bytes
Save(<data/9c119f1243>) returned error, retrying after 465.077794ms: wrote 0 bytes instead of the expected 5340921 bytes
Save(<data/26c2699481>) returned error, retrying after 573.840022ms: wrote 0 bytes instead of the expected 4836380 bytes
Save(<data/19e3a667a6>) returned error, retrying after 386.829792ms: wrote 0 bytes instead of the expected 4285304 bytes
Save(<data/9ec3bc2621>) returned error, retrying after 372.787189ms: wrote 0 bytes instead of the expected 4816941 bytes
Save(<data/126fb66282>) returned error, retrying after 407.907788ms: wrote 0 bytes instead of the expected 8044893 bytes
Save(<data/b809013493>) returned error, retrying after 585.885394ms: wrote 0 bytes instead of the expected 5019476 bytes
Save(<data/da4f156b6e>) returned error, retrying after 399.784602ms: wrote 0 bytes instead of the expected 4251178 bytes
Save(<data/ce8189faac>) returned error, retrying after 359.491966ms: wrote 0 bytes instead of the expected 7606047 bytes
Save(<data/1a7b70e0a9>) returned error, retrying after 489.422546ms: wrote 0 bytes instead of the expected 6222493 bytes
Save(<data/ac5cd08951>) returned error, retrying after 282.99054ms: wrote 0 bytes instead of the expected 4737552 bytes
Save(<data/dd0372bb57>) returned error, retrying after 702.726508ms: wrote 0 bytes instead of the expected 7334049 bytes
Save(<data/67f3ec2c46>) returned error, retrying after 316.214175ms: wrote 0 bytes instead of the expected 5088447 bytes
Save(<lock/54cc7a7b8a>) returned error, retrying after 354.556843ms: wrote 0 bytes instead of the expected 153 bytes
[58:50:48] 29.33%  6538380 files 4.072 TiB, total 19351732 files 13.883 TiB, 0 errors ETA 141:48:02
[ backup continues ]
```

```
restic version
restic 0.12.0 compiled with go1.15.8 on linux/amd64
```

When backup takes a long time it really makes sense to have a more detailed trace from around when problems are being reported, along with a confirmation by stat that the actions succeeded. Even listing files which were processed when fault handling is activated could help - one could simply attempt a targeted selective recovery on them once the job finishes

**Comment by MichaelEischer** [comment]
> When backup takes a long time it really makes sense to have a more detailed trace from around when problems are being reported, along with a confirmation by stat that the actions succeeded. Even listing files which were processed when fault handling is activated could help - one could simply attempt a targeted selective recovery on them once the job finishes

Enabling a trace on demand  is not that simple. When the first error shows up then, the original request has already been sent and won't be logged unless the debug log is collected all the time.

The fault handling happens per file, so the log snippet already contains a list of all affected files. However, using that information later on to try to repair a backup sounds like a horrible hack. restic has to rely on a backend to properly store uploaded files or otherwise all sorts of strange behaviors might show up as this breaks some assumptions made by the code.

I've created an experimental branch https://github.com/restic/restic/tree/debug-stat-on-retry (based on restic 0.12.0) which will call stat for retried uploads. The branch will also print the result of the delete operation if that resulted in an error. Could you test that branch?

**Comment by borkd** [comment]
Thanks @MichaelEischer. Building the binary for https://github.com/restic/restic/tree/debug-stat-on-retry now, will use it for some of the larger (re)runs

**Comment by borkd** [comment]
@MichaelEischer - I've been using the stat-on-retry branch since and cannot report any adverse effects of having that feature available as an option in the mainstream branch. Will provide log output when I run into the backend having hiccups again

**Comment by etheling** [comment]
I think this is sort of related to the issue here - my apologies if I should have created a new ticket instead.

I am using Backblaze B2 as backend and similar to the problem described in the ticket, I had a long running backup job that exceed my quota. Restic has been trying for days to continue failing backup. Desired behavior would have been restic to exit with an error code after some amount of retries.

`
Save(<data/011a19a1fa>) returned error, retrying after 552.330144ms: Close: b2_upload_file: 403: Cannot upload files, storage cap exceeded. See the Caps & Alerts page to increase your cap.
`

What lead to this is that [accidentally] some large directories were copied to the backed up drive. I did eventually get an alert from B2 that I've hit my quota limits and realized what was going on (and killed ongoing never ending restic backup). My recovery plan was to move the large directories that ended up being backed up somewhere else, list snapshots, then delete/forget the latest snapshot and purge the files. Alas, as I am already at my quota, all I am getting is:

```
root@abdf97fc94a2:/opt/restic/bin# ./restic snapshots
repository 9c4888b4 opened successfully, password is correct
Save(<lock/893617a1d0>) returned error, retrying after 552.330144ms: Close: b2_get_upload_url: 403: Cannot upload files, storage cap exceeded. See the Caps & Alerts page to increase your cap.

```
----

```
root@abdf97fc94a2:/opt/restic/bin# /opt/restic/bin/restic forget --keep-within-daily 7d --keep-within-weekly 1m --keep-within-monthly 1y --keep-within-yearly 1y
repository 9c4888b4 opened successfully, password is correct
Save(<lock/f0aab7aa8e>) returned error, retrying after 552.330144ms: Close: b2_get_upload_url: 403: Cannot upload files, storage cap exceeded. See the Caps & Alerts page to increase your cap.
```

or even if I knew the snapshot id

```
root@abdf97fc94a2:/opt/restic/bin# /opt/restic/bin/restic forget a5596887
repository 9c4888b4 opened successfully, password is correct
Save(<lock/86b4a9932d>) returned error, retrying after 552.330144ms: Close: b2_get_upload_url: 403: Cannot upload files, storage cap exceeded. See the Caps & Alerts page to increase your cap.
```


---

E.g. I believe I am stuck between: increasing my caps to be able to purge the latest snapshot, or deleting entire B2 bucket with incremental backups in it and starting all over.

I am not familiar with B2 API, so maybe part of problem lies with it - e.g. in order to list snapshots or remove files, it is required to issue API calls that are blocked because of exceeded quota. Either way, just wanted to chime in regarding this as I think the fact that restic wont exit after certain amount of errors, followed by inability to communicate to B2 create a somewhat troublesome situation that - if possible - could be handled more gracefully.

**Comment by borkd** [comment]
Was not able to reproduce this problem outside of the swiftstack's s3. Closing
